reader_concurrency_semaphore.cc

```diff
@@ -326,6 +327,22 @@ class reader_permit::impl
     void set_max_result_size(query::max_result_size s) {
         _max_result_size = std::move(s);
     }
+
+    void on_start_sstable_read() noexcept {
+        if (!_sstables_read) {
+            ++_semaphore._stats.disk_reads;
+        }
+        ++_sstables_read;
+        ++_semaphore._stats.sstables_read;
+    }
+
+    void on_finish_sstable_read() noexcept {
+        --_sstables_read;
+        --_semaphore._stats.sstables_read;
+        if (!_sstables_read) {
+            --_semaphore._stats.disk_reads;
+        }
+    }
```

Comment:
We can make it safer, without trusting the user, that it will always call on_finish_sstable_read() even on failure.

What about we add this to dtor:

_semaphore._stats.sstables_read -= _sstables_read;
_semaphore._stats.disk_reads -= bool(_sstables_read);

?

Reply:
This would substract all reads at once, which would obscure the metrics reads that release sstable readers incrementally.

Reply:
But the reader_permit outlives the readers, right? reader_permit is responsible for managing the stats, therefore it should avoid any leaks once destroyed. subtracting all reads added by a given permit at once is better than never subtracting them at all.

Reply:
Ah you mean to prevent leaks. I suppose we can do that, we already do it for some other counters.

Reply:
Yes :-) Not a replacement to what you implemented, but just a defense mechanism.